src/librbd/crypto/BlockCrypto.cc

```diff
@@ -90,25 +74,47 @@ int BlockCrypto<T>::crypt(ceph::bufferlist* data, uint64_t image_offset,
         remaining_buf_bytes -= copy_size;
       }
 
-      int crypto_output_length = 0;
+      const unsigned char* crypto_in_ptr = nullptr;
       if (leftover_size == 0) {
-        crypto_output_length = m_data_cryptor->update_context(
-              ctx, in_buf_ptr, out_buf_ptr, m_block_size);
-
+        crypto_in_ptr = in_buf_ptr;
         in_buf_ptr += m_block_size;
         remaining_buf_bytes -= m_block_size;
       } else if (leftover_size == m_block_size) {
-        crypto_output_length = m_data_cryptor->update_context(
-              ctx, leftover_block, out_buf_ptr, m_block_size);
+        crypto_in_ptr = leftover_block;
         leftover_size = 0;
+      } else {
+        continue;
       }
 
-      if (crypto_output_length < 0) {
-        lderr(m_cct) << "crypt update failed" << dendl;
-        return crypto_output_length;
+      unsigned char* out_buf_ptr = reinterpret_cast<unsigned char*>(
+              appender.get_pos_add(m_block_size));
+
+      if (mode == CIPHER_MODE_DEC &&
+        mem_is_zero(reinterpret_cast<const char*>(crypto_in_ptr),
+                    m_block_size)) {
+        // input is already plaintext (zeros), so don't decrypt
+        memset(out_buf_ptr, 0, m_block_size);
```

Comment:
Wouldn't this result in corrupt data being returned to the user for a bona fide all-zeroes ciphertext block?  Is there something I'm missing that makes all-zeroes ciphertext blocks impossible here?

Because a block cipher, under a given key, is supposed to be a given bijective (one-to-one) mapping on all possible n-bit strings (where n is block size), meaning that there is always going to be some plaintext bit string corresponding to all-zeroes ciphertext bit string.

Reply:
I think this should be taken in the opposite direction -- instead of returning zeroes to the user if the object doesn't exist (OSD returns `ENOENT`), these zeroes should be run through the decryption.

Reply:
> Wouldn't this result in corrupt data being returned to the user for a bona fide all-zeroes ciphertext block? Is there something I'm missing that makes all-zeroes ciphertext blocks impossible here?
> 
> Because a block cipher, under a given key, is supposed to be a given bijective (one-to-one) mapping on all possible n-bit strings (where `n` is block size), meaning that there is always going to be some plaintext bit string corresponding to all-zeroes ciphertext bit string.

From a theoretical point of view you're right.
But practically, the probability for that edge-case is 2 ^ block_size, i.e. at most 2 ^ -512.
This probability is so small, that is considered practically impossible. Personally, I will not fear to take a chance on my own life with this probability. But perhaps this is more of a philosophical debate.

Anyhow, if you don't take that chance, the user "loses" this initialization convention which exists for regular (non-encrypted) images.

Reply:
> From a theoretical point of view you're right. But practically, the probability for that edge-case is 2 ^ block_size_in_bits, i.e. at most 2 ^ -4096. This probability is so small, that is considered practically impossible. Personally, I will not fear to take a chance on my own life with this probability. But perhaps this is more of a philosophical debate.

Are you aware of any storage system with encryption support doing that?

> 
> Anyhow, if you don't take that chance, the user "loses" this initialization convention which exists for regular (non-encrypted) images.

That is fine.  There is no way to make regular and formatted images behave the same in all aspects anyway.
Meanwhile, the existing behavior means that even flat images or simple clones aren't actually compatible with LUKS despite the "fully compliant with the LUKS specification" claim.

Reply:
> 
> Are you aware of any storage system with encryption support doing that?
> 

@dannyharnik can you please give an example?

> 
> That is fine. There is no way to make regular and formatted images behave the same in all aspects anyway. Meanwhile, the existing behavior means that even flat images or simple clones aren't actually compatible with LUKS despite the "fully compliant with the LUKS specification" claim.

I see at least 2 issues with keeping the current "allow decrypting zeros" logic:
1. A user can get non-consistent data read from librbd, depending on the block size of its reads.
2. qemu assumes all rbd images are zero initialized, though this can be fixed by issuing a patch to qemu rbd driver. see this line here: https://github.com/qemu/qemu/blob/master/block/rbd.c#L1633

Reply:
> >  Are you aware of any storage system with encryption support doing that?

> @dannyharnik can you please give an example

I've seen such designs internally of encryption in storage systems, but these are not public and I can't give you a pointer. 
Similarly there are other designs of storage systems that allow a low probability of data corruption that is purely statistical as long as it is small enough. The main example here is with deduplication, where it is common practice to store data solely according to its hash, and a hash collision would therefore entail data loss. Data domain have publicly acknowledged this (see link below) , as did IBM storage. In these cases a hash of 160 bits is considered sufficiently long, whereas the case we are dealing with much lower probability of at least 4096 bits, which is an entirely negligible event.

Data Domain's paper - https://www.usenix.org/legacy/event/fast08/tech/full_papers/zhu/zhu.pdf
The relevant part is section 2.4

Reply:
I was hoping to see a reference to a well-known storage system doing this specifically in its encryption layer.  Data deduplication isn't exactly the same scenario.  Data Domain notwithstanding, many data deduplication solutions provide a "verify" (paranoid mode) configuration option that employs a byte-by-byte comparison if the hashes match, no matter how strong the hash is -- and there are still endless arguments on that...

In a data deduplication scenario, the benefit is clear -- skipping the potential read from the backing store + byte comparison makes it faster.  I'm not sure I understand the benefits of introducing this "don't decrypt zeroes" special case into a full-disk encryption scheme.  Other than something like "qemu-img convert" being able to assume that the image it created is zeroed and skip writing zeroes, what else?

Reply:
> I see at least 2 issues with keeping the current "allow decrypting zeros" logic:
> 
> 1. A user can get non-consistent data read from librbd, depending on the block size of its reads.

Did you see my earlier comment on resolving this inconsistency by *always* decrypting synthesized zeroes before returning them to the user?  I can't think of any inconsistencies there as we would no longer depend on sparse-read OSD op.

> 2. qemu assumes all rbd images are zero initialized, though this can be fixed by issuing a patch to qemu rbd driver. see this line here: https://github.com/qemu/qemu/blob/master/block/rbd.c#L1633

Right.  Note that the generic `luks` driver in QEMU doesn't provide `bdrv_has_zero_init` precisely because zeroes are always decrypted in LUKS.

Reply:
The issue is the interaction between thin provisioned disks and encryption. It is pretty clear that there is no functional benefit in returning garbage from non-allocated areas. This is just how its been done when combining a thin-provisioned storage with an encryption layer that was built for a general disk format. I also see no security benefit in doing so. 
If you integrate a thin-provisioned format with encryption (like in qcow2-LUKS) you can do things differently.  
 
The main benefit of not decrypting zeros is to handle operations like migration or re-encryption (either to change keys or change format) without losing the thin provisioning. I think these are significant.

Reply:
> The issue is the interaction between thin provisioned disks and encryption. It is pretty clear that there is no functional benefit in returning garbage from non-allocated areas. This is just how its been done when combining a thin-provisioned storage with an encryption layer that was built for a general disk format. I also see no security benefit in doing so. If you integrate a thin-provisioned format with encryption (like in qcow2-LUKS) you can do things differently.

I mostly agree and I'm aware that qcow2-luks indeed does it differently, even though it still calls itself luks.  And I don't have any objections to doing it the qcow2-luks way -- note that qcow2-luks doesn't need to resort to "let's skip decrypting zeroes" or any other hack^Wshortcuts to achieve that.  It just falls naturally from the fact that it always writes entire clusters (without extended L2) and that the sub-cluster allocation information is always available (with extended L2).

The problem with doing something similar here is that there doesn't seem to be a reliable way of getting the sub-object extent map in Ceph.  The sparse-read op just outright doesn't work on EC pools and the guarantees for replicated pools aren't clear either.  I have started a thread to get that clarified here:

https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/E7B62CQKWJ3QL2QMQJ43SQVL4LVVTTEM/

(I CCed Or, should have CCed you as well, sorry!)

> 
> The main benefit of not decrypting zeros is to handle operations like migration or re-encryption (either to change keys or change format) without losing the thin provisioning. I think these are significant.

I think we can still do that (sparse copyup and therefore flatten, migration, etc) with decrypting zeroes.  The issue of how synthesized zeroes are presented to the user and the issue of maintaining sparseness when possible (i.e. when sparse-read does the right thing) seem completely orthogonal to me.

Reply:
> The issue of how synthesized zeroes are presented to the user and the issue of maintaining sparseness when possible (i.e. when sparse-read does the right thing) seem completely orthogonal to me.

I agree on this. What a user is presented is not actually important to any use-case I am aware of. The only case would be an application that explicitly expects zeros on non-initialized areas,  but such an application would fail with most of today's encryption implementations.

As for sparseness, there are other workarounds. This would have made life simpler...